### DISCUSSION TITLE
Rotations along Member element

### DISCUSSION THREAD

**Discussion by roger1017** [JWock82/Pynite#159] [Q&A] [answered]
Hi,

Could anyone show me how to get rotation results out from the model? I am interested in rotation results between two nodes.

Thanks!

BR
Roger

**Comment by JWock82** [comment]
There are a couple ways to do it, but it's a bit tedious right now. I have not built out a "slope" method for the `Member3D` object like I have for deflection.

You could use the `deflection` method as a template to create a `slope` method in the `Member3D` object. There are already slope methods in the underlying beam segment classes that are used to build member results. The math is already done, it just hasn't been presented for the user in a clean easy to use wrapper function like deflections have.

Or, another way is to get the rotation at the start of the member, and then add the result of numerically integrating the moment diagram divided it by E*I for the member up to the point of interest. `Scipy` does a great job at numerically integrating python functions.

**Reply by roger1017** [reply]
Hi Craig,

Many thanks for the quick reply. 

I did try to use the build in slope function to get rotation outputs. But I got slightly different results (0.1%) compared with the rotation I got directly from Node.RX. When I use deflection() I got exact the same results. Also what would you recommend for the torsional displacement calculation? 

   def rotation(self, Direction, x, combo_name="Combo 1"):
        """
        Returns the rotation at a point along the member's length.

        Parameters
        ----------
        Direction : string
            The direction in which to find the deflection. Must be one of the following:
                'ry' = Deflection in the local y-axis.
                'rz' = Deflection in the local z-axis.
        x : number
            The location at which to find the rotation.
        combo_name : string
            The name of the load combination to get the results for (not the load combination itself).
        """

        # Segment the member if necessary
        if self.__solved_combo == None or combo_name != self.__solved_combo.name:
            self._segment_member(combo_name)
            self.__solved_combo = self.LoadCombos[combo_name]

        # Check which axis is of interest
        if Direction == "rx":

            raise ValueError('Torsional displacement not allowed at this moment')

        elif Direction == "ry":

            # Check which segment 'x' falls on
            for segment in self.SegmentsZ:

                if round(x, 10) >= round(segment.x1, 10) and round(x, 10) < round(segment.x2, 10):

                    return segment.Slope(x - segment.x1)

            if isclose(x, self.L()):

                lastIndex = len(self.SegmentsZ) - 1
                return self.SegmentsZ[lastIndex].Slope(x - self.SegmentsZ[lastIndex].x1)

        elif Direction == "rz":

            for segment in self.SegmentsY:

                if round(x, 10) >= round(segment.x1, 10) and round(x, 10) < round(segment.x2, 10):

                    return segment.Slope(x - segment.x1)

            if isclose(x, self.L()):

                lastIndex = len(self.SegmentsY) - 1
                return self.SegmentsY[lastIndex].Slope(x - self.SegmentsY[lastIndex].x1)

**Comment by JWock82** [comment, marked as answer]
Without diving too deep into it, that method you wrote looks about right. 0.1% sounds pretty accurate. Keep in mind these slopes are measured in radians, and the numbers are very small, so we may be starting to butt up against precision tracking limits in python. Torque is a different animal altogether. There are no built-in functions in the segment class to help you with that. You'd have to make some assumptions about the internal rotation as a function of loads and end rotations. That depends on the shape of the cross section. AISC Design Guide 9 is a testament to how complex those relationships can be. I was always taught that in design the first rule of torsion is to eliminate it.

**Reply by roger1017** [reply]
Thanks for the answer!
